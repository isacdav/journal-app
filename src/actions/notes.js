import Swal from 'sweetalert2';
import { db } from '../firebase/firebase';
import { fileUpload } from '../helpers/fileHelper';
import { loadNotes } from '../helpers/notesHelper';
import { types } from '../types/types';

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime(),
    };

    const doc = await db.collection(`${uid}/journal/notes`).add(newNote);

    dispatch(activateNote(doc.id, newNote));
    dispatch(addNewNote(doc.id, newNote));
  };
};

export const addNewNote = (id, note) => ({
  type: types.notesAddNew,
  payload: {
    id,
    ...note,
  },
});

export const startLoadingNotes = uid => {
  return async dispatch => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const activateNote = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note,
  },
});

export const setNotes = notes => ({
  type: types.notesLoad,
  payload: notes,
});

export const startSaveNote = note => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    if (!note.url) delete note.url;

    const noteToDb = { ...note };
    delete noteToDb.id;

    await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToDb);

    dispatch(refreshNote(note.id, noteToDb));

    Swal.fire('Saved', note.title, 'success');
  };
};

export const refreshNote = (id, note) => ({
  type: types.notesUpdated,
  payload: {
    id,
    note: {
      id,
      ...note,
    },
  },
});

export const startUploadingFile = file => {
  return async (dispatch, getState) => {
    const { active } = getState().notes;

    Swal.fire({
      title: 'Uploading...',
      text: 'Please wait while the picture is being uploaded.',
      allowOutsideClick: false,
      willOpen: () => {
        Swal.showLoading();
      },
    });

    const fileUrl = await fileUpload(file);
    active.url = fileUrl;

    dispatch(startSaveNote(active));

    Swal.close();
  };
};

export const startDeleting = id => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;
    await db.doc(`${uid}/journal/notes/${id}`).delete();

    dispatch(deleteNote(id));
  };
};

export const deleteNote = id => ({
  type: types.notesDelete,
  payload: id,
});

export const cleanNotesLogout = () => ({
  type: types.notesLogoutCleaning,
});
