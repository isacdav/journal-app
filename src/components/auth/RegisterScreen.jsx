import React from 'react';
import { Link } from 'react-router-dom';

const RegisterScreen = () => {
  return (
    <>
      <h2 className='auth__title mb-20'>Hello</h2>
      <form>
        <input
          className='auth__input'
          type='text'
          placeholder='Name'
          name='name'
          autoComplete='off'
        />
        <input
          className='auth__input'
          type='email'
          placeholder='Email'
          name='email'
          autoComplete='off'
        />
        <input
          className='auth__input'
          type='password'
          placeholder='Password'
          name='password'
        />
        <input
          className='auth__input'
          type='password'
          placeholder='Confirm password'
          name='password2'
        />
        <button className='btn btn-primary btn-block mb-5' type='submit'>
          Log In
        </button>
        <small>
          <Link className='link' to='/auth/login'>
            Already registered?
          </Link>
        </small>
      </form>
    </>
  );
};

export default RegisterScreen;
