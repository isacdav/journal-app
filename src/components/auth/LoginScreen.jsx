import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { startGoogleLogin, startLoginEmail } from '../../actions/auth';
import { removeError, setError } from '../../actions/ui';
import useForm from '../../hooks/useForm';

const LoginScreen = () => {
  const dispatch = useDispatch();
  const { msgError, loading } = useSelector(state => state.ui);

  const [formValues, handleInputChange] = useForm({
    email: '',
    password: '',
  });

  const { email, password } = formValues;

  const handleSubmit = e => {
    e.preventDefault();

    if (!isFormValid()) return;

    dispatch(startLoginEmail(email, password));
  };

  const isFormValid = () => {
    if (email.trim().length <= 0 || password.trim().length <= 0) {
      dispatch(setError('Please enter Email and Password'));
      return false;
    }

    dispatch(removeError());
    return true;
  };

  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  };

  return (
    <>
      <h2 className='auth__title mb-20'>Hello</h2>
      <form
        onSubmit={handleSubmit}
        className='animate__animated animate__fadeIn animate__faster'>
        {msgError && <div className='auth__alert-error'>{msgError}</div>}
        <input
          className='auth__input'
          id='input_email'
          type='email'
          placeholder='Email'
          name='email'
          autoComplete='off'
          value={email}
          onChange={handleInputChange}
        />
        <input
          className='auth__input'
          id='input_pass'
          type='password'
          placeholder='Password'
          name='password'
          value={password}
          onChange={handleInputChange}
        />
        <button
          className='btn btn-primary btn-block mb-5'
          type='submit'
          disabled={loading}>
          Log In
        </button>
        <small>
          <Link className='link' to='/auth/register'>
            New account
          </Link>
        </small>

        <div className='auth__social-net mt-20'>
          <div className='google-btn' onClick={handleGoogleLogin}>
            <div className='google-icon-wrapper'>
              <img
                className='google-icon'
                src='https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg'
                alt='google button'
              />
            </div>
            <p className='btn-text'>
              <b>Sign in with Google</b>
            </p>
          </div>
        </div>
      </form>
    </>
  );
};

export default LoginScreen;
