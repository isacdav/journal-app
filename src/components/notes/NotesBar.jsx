import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUploadingFile } from '../../actions/notes';

const NotesBar = () => {
  const dispatch = useDispatch();

  const { active } = useSelector(state => state.notes);

  const handleSave = () => {
    dispatch(startSaveNote(active));
  };

  const handlePictureClick = () => {
    document.querySelector('#file_selector').click();
  };

  const handleFileChange = e => {
    const file = e.target.files[0];
    if (file) {
      dispatch(startUploadingFile(file));
    }
  };

  const currentDate = new Date().toDateString();

  return (
    <div className='notes__bar'>
      <span>{currentDate}</span>
      <input
        id='file_selector'
        type='file'
        style={{ display: 'none' }}
        onChange={handleFileChange}
      />
      <div>
        <button className='btn' onClick={handlePictureClick}>
          Picture
        </button>
        <button className='btn' onClick={handleSave}>
          Save
        </button>
      </div>
    </div>
  );
};

export default NotesBar;
