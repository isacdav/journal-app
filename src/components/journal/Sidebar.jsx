import React from 'react';
import { useDispatch } from 'react-redux';
import { startLogout } from '../../actions/auth';
import JournalEntries from './JournalEntries';

const Sidebar = () => {
  const dispatch = useDispatch();

  const handleLogout = () => {
    dispatch(startLogout());
  };

  return (
    <aside className='journal__sidebar'>
      <div className='journal__sidebar-nav'>
        <h3 className='mt-20'>
          <i className='far fa-moon'></i>
          <span> Isaac</span>
        </h3>

        <button className='btn' onClick={handleLogout}>
          Logout
        </button>
      </div>

      <div className='journal__new'>
        <i className='far fa-calendar-plus fa-4x'></i>
        <p className='mt-5'>New Entry</p>
      </div>

      <JournalEntries />
    </aside>
  );
};

export default Sidebar;
