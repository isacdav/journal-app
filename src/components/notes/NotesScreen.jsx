import React, { useRef, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { activateNote, startDeleting } from '../../actions/notes';
import useForm from '../../hooks/useForm';
import NotesBar from './NotesBar';

const NotesScreen = () => {
  const dispatch = useDispatch();

  const { active: note } = useSelector(state => state.notes);
  const [formValues, handleInputChange, reset] = useForm(note);
  const { title, body, id } = formValues;

  // Save the current id to compare later
  const activeId = useRef(note.id);

  const handleDelete = () => {
    dispatch(startDeleting(id));
  };

  useEffect(() => {
    // Validate if the note has changed
    if (note.id !== activeId.current) {
      reset(note);
      activeId.current = note.id;
    }
  }, [note, reset]);

  useEffect(() => {
    dispatch(activateNote(formValues.id, { ...formValues }));
  }, [formValues, dispatch]);

  return (
    <div className='notes__main'>
      <NotesBar />
      <div className='notes__content'>
        <input
          type='text'
          name='title'
          placeholder='My title'
          className='notes__title'
          autoComplete='off'
          value={title}
          onChange={handleInputChange}
        />
        <textarea
          name='body'
          placeholder='A description of my day...'
          className='notes__textarea'
          value={body}
          onChange={handleInputChange}></textarea>
        {note.url && (
          <div className='notes__image'>
            <img src={note.url} alt='My day photography' />
          </div>
        )}
      </div>

      <button className='btn btn-danger' onClick={handleDelete}>
        Delete
      </button>
    </div>
  );
};

export default NotesScreen;
