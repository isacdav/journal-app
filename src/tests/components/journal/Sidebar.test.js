import '@testing-library/jest-dom';
import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Sidebar from '../../../components/journal/Sidebar';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { startLogout } from '../../../actions/auth';
import { startNewNote } from '../../../actions/notes';

jest.mock('../../../actions/auth', () => ({
  startLogout: jest.fn(),
}));

jest.mock('../../../actions/notes', () => ({
  startNewNote: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
  auth: {
    uid: 'ahwoencwecnowe',
    name: 'El Patron',
  },
  ui: {
    loading: false,
    msgError: null,
  },
  notes: {
    active: null,
    notes: [],
  },
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <Sidebar />
  </Provider>
);

describe('<Sidebar /> Tests', () => {
  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should call logout', () => {
    wrapper.find('button').prop('onClick')();
    expect(startLogout).toHaveBeenCalled();
  });

  test('should call startNewNote', () => {
    wrapper.find('.journal__new').prop('onClick')();
    expect(startNewNote).toHaveBeenCalled();
  });
});
