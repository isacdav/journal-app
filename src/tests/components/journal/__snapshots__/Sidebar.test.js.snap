// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> Tests should match snapshot 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-nav"
  >
    <h3
      className="mt-20"
    >
      <i
        className="far fa-moon"
      />
      <span>
         
        El Patron
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-4x"
    />
    <p
      className="mt-5"
    >
      New Entry
    </p>
  </div>
  <div
    className="journal__entries"
  />
</aside>
`;
